internal/pkg/nats/client.go

```diff
@@ -0,0 +1,151 @@
+//
+// Copyright (c) 2022 One Track Consulting
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//go:build include_nats_messaging
+
+package nats
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg/nats/interfaces"
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
+	"github.com/nats-io/nats.go"
+)
+
+func NewConnection(cc ClientConfig) (interfaces.Connection, error) {
+	nc, err := nats.Connect(cc.BrokerURL, cc.ConnectOpt()...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return nc, nil
+}
+
+// NewClient initializes creates a new client using NATS core
+func NewClient(cfg types.MessageBusConfig) (*Client, error) {
+	return NewClientWithConnectionFactory(cfg, NewConnection)
+}
+
+// NewClientWithConnectionFactory creates a new client that uses the specified function to establish pub/sub connections
+func NewClientWithConnectionFactory(cfg types.MessageBusConfig, connectionFactory ConnectNats) (*Client, error) {
+	if connectionFactory == nil {
+		return nil, fmt.Errorf("connectionFactory is required")
+	}
+
+	var m interfaces.MarshallerUnmarshaller = &edgexMarshaller{} // default
+
+	cc, err := CreateClientConfiguration(cfg)
+
+	if err != nil {
+		return nil, err
+	}
+
+	switch strings.ToLower(cc.Format) {
+	case "nats":
+		m = &natsMarshaller{}
+	case "json":
+		m = &jsonMarshaller{}
+	case "cbor":
+		m = &cborMarshaller{}
+	}
```

Comment:
One argument for the other marshalers could be if someone was using the MQTT support (https://docs.nats.io/running-a-nats-service/configuration/mqtt ) to connect legacy services.  Since the other services would need to be able to speak in JSON/CBOR envelopes over MQTT it could make sense to use the other format to enable NATS throughout as an alternative to some kind of adapter service that handles the encoding change.  I wonder if nats should be default though 🤔

Reply:
Since this is for our internal EdgeX MessageBus, the `connect legacy services` scenario is out of scope. All Edgex Services (that use the MessageBus) must be configured to use the same MessageBus type (redis, mqtt, zero, or nats). 

Also, it is only the payload this is either JSON or CBOR encode by the client and passed in via the MessageEnvelope. The MessageEnvelope JSON encode in the other implementations, which results in the payload being double encoded. 

Lets keep is simple and take advantage of NATs capability to avoid the double encoding of the payload like we do with the other implementations.

Reply:
Ok, I now see the need to have the two formats 

- "nats" for when all service have messagebus config set to `nats` 
- "edgex" for when other services have messagebus config set to `mqtt` and the broker is the NATS server in MQTT mode.
   - Note that this format is setting the NATs message to be the JSON encoded MessageEnvelope just like the other implementation do.